- challenge table

ssardana08 opened the pull request:
## Changes

## Fixes


## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

ssardana08 committed:
- challenge table 1st commit

ssardana08 committed:
- name correct for LatestCompletedTxn
- methods added

ssardana08 committed:
- allocation challenge removed

ssardana08 committed:
- clean

ssardana08 committed:
- cancelledPassRate fixed with blobberchallenge

ssardana08 committed:
- bug fix on index iteration

ssardana08 committed:
- non expired challenges left for cancelled allocation should not be considered as successful

ssardana08 committed:
- blobberchallenge rightly populated

ssardana08 committed:
- logic correction

ssardana08 committed:
Merge branch 'staging' into challenges_table

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -1235,38 +1235,43 @@ func (sc *StorageSmartContract) canceledPassRates(alloc *StorageAllocation,
+		case nil:
+			for i := 0; i < len(bc.Challenges); i++ {
+				chall := bc.Challenges[i]
+				if chall.AllocationID == alloc.ID {
```
We only need to check challenges that has expired, since all challenges are already sorted by timestamp in asc, so we can break once detected a challenge that is not expired. The persudo code could be:

```go
for i := 0; i < len(challenges); i++ {
    clg := challenges[i]
    if !isExpired(clg, now) {
        break
    }
    
    if clg.AllocationID != alloc.ID {
        continue
    }

    // found the allocation 
    //  remove the challenge and update stats...
}
```

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -101,11 +101,17 @@ type ChallengeResponse struct {
+type BlobberChallengeData struct {
+	ID           string           `json:"id"`
+	AllocationID string           `json:"allocation_id"`
+	Created      common.Timestamp `json:"created"`
```
CratedAt would be better, Created sounds like a bool variable.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -131,8 +137,8 @@ func (sn *BlobberChallenge) Decode(input []byte) error {
 		return err
 	}
 	sn.ChallengeIDMap = make(map[string]struct{})
```
Initialize the map with capacity

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -154,77 +160,53 @@ func (sn *BlobberChallenge) UnmarshalMsg(data []byte) ([]byte, error) {
 	*sn = BlobberChallenge(*d)
 
 	sn.ChallengeIDMap = make(map[string]struct{})
```
Init the map with capacity

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -154,77 +160,53 @@ func (sn *BlobberChallenge) UnmarshalMsg(data []byte) ([]byte, error) {
+		}
+		delete(sn.ChallengeIDMap, chall.ID)
+		alloc.Stats.ChallengeFailed(chall.ID)
+		alloc.BlobberMap[sn.BlobberID].Stats.ChallengeFailed(chall.ID)
```
The challenges could be for other allocations.

ssardana08 committed:
- removed Blobbers from storageAllocation
- added new struct AllocationInfo, which is used by API with blobbers

ssardana08 committed:
- handler response

ssardana08 committed:
- removing expired allocation challenges

ssardana08 committed:
- CreatedAt property name updated

ssardana08 committed:
- cancelAllocation to get pass rates as finalizedPassRates

ssardana08 committed:
- cancelAllocation to get pass rates as finalizedPassRates

ssardana08 committed:
Merge branch 'staging' into challenges_table

ssardana08 committed:
- cancelledPassRates updated

ssardana08 committed:
- challenge table implemented

ssardana08 committed:
- prev id fixed

ssardana08 committed:
Merge branch 'staging' into challenges_table

ssardana08 committed:
- msgpack updated

ssardana08 committed:
- challenge info

ssardana08 committed:
- challenge info added

ssardana08 committed:
- blobber challenge removed
- challenges moved to allocation object
- blobber storing latest openchallengeid and latestcompletedchallenge

ssardana08 committed:
- cancelAllocation remove expired challenge first and then mark remaining success

ssardana08 committed:
- remove expired challenges fixed

ssardana08 committed:
- lint errors

Sriep commented on code/go/0chain.net/smartcontract/storagesc/blobber.go:
```diff
@@ -52,6 +52,22 @@ func (sc *StorageSmartContract) getBlobber(blobberID string,
 	return
 }
 
+func (sc *StorageSmartContract) getBlobbers(blobberIDs []string,
```
You could replace with [getAllocationBlobbers](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/allocation.go#L525) which seems to do the same thing, but more efficiently. Or remove getAllocationBlobbers, and redo `getBlobbers` in the same syle as getAllocationBlobbers.

Sriep commented on code/go/0chain.net/smartcontract/dbs/event/challenge.go:
```diff
@@ -0,0 +1,114 @@
+	return result.Error
+}
+
+func (edb *EventDb) addOrOverwriteChallenge(ch *Challenge) error {
```
If we are never going to overwrite a challenge then maybe we should change this to `addChallenge` and raise an error if it already exists.

Sriep commented on code/go/0chain.net/smartcontract/dbs/event/process.go:
```diff
@@ -50,6 +51,8 @@ const (
 	TagStakePoolReward
 	TagUpdateDelegatePool
 	TagAddOrOverwriteAllocation
+	TagAddOrOverwriteChallenge
```
Suggest replace with `TagAddChallenge`

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go:
```diff
@@ -0,0 +1,94 @@
+	}
+}
+
+func emitAddOrOverwriteChallenge(ch *StorageChallengeInfo, balances cstate.StateContextI) error {
```
If we never intend to overwrite a challenge then I suggest renaming to emitAddChallenge.

Sriep reviewed (commented):
What are we doing about challenge completion time?
